### ISSUE TITLE
Retain additional adapter information (timing, timeout, etc)

### ISSUE THREAD

**Issue by nevans** [lostisland/faraday#527]
Some adapters (e.g. typhoeus) have detailed timing information (`total_time`, `starttransfer_time`, `appconnect_time`, `pretransfer_time`, `connect_time`, `namelookup_time`, etc) that they may have gotten from the underlying C library, so discovering timing information via middleware (e.g. `start = Time.now; app.call(env).on_complete do |env| duration = Time.now - start end`) has an inherent loss of fidelity.

Also, when operating in parallel, adapters don't/can't/shouldn't raise exceptions on timeouts (since we don't currently have any mechanism for handling `on_complete` callback exceptions when running in parallel).  One hack for this is to check for `response.status == 0`, but it would be better to just ask `response.timed_out?` (there may be other bugs that could lead to `status == 0`) and it's unhelpful to conflate them.  We ought to preserve this information on the response.

````ruby
response.timed_out? #=> true/false
response.time       #=> total time in seconds
response.time_ms    #=> total time in ms
response.times      #=> {total: t0, starttransfer: t1, appconnect: t2, ...}
response.times      #=> {total: t0, starttransfer: t1, appconnect: t2, ...}
response.adapter    #=> whatever other adapter specific response info, perhaps nothing
```

Stored on the response under the obvious keys: `:timed_out`, `:times_ms`, and `:adapter`.  The catch-all is because it's still better to give a place to put other info than to lose it entirely.  However, we could also just toss those adapter specific odds-and-ends onto a `custom_member` of `env` (see also #525).

I already have a PR at typhoeus/typhoeus#476 to update the typhoeus adapter with `env[:typhoeus_timed_out]`, but it would make more sense to have a more standard mechanism that could be adopted by any adapters.  If you think this is a reasonable approach (or have a different approach to suggest), I'll try my hand at updating some of other adapters to add this info.

**Comment by nevans** [comment]
BTW, I notice that [hurley](https://github.com/lostisland/hurley/blob/3d3b11fdaeea7107d00b448edbd08fa452d524e5/lib/hurley/client.rb#L250-L258) already has [response.ms](https://github.com/lostisland/hurley/blob/3d3b11fdaeea7107d00b448edbd08fa452d524e5/lib/hurley/client.rb#L317-L319) as part of its API, although it's getting it by measuring `Time.now` since `Net::HTTP` doesn't report it.  But I won't be converting to hurley any time soon (not least because it doesn't have parallel support... yet?).

**Comment by nevans** [comment]
FYI: I'm just looking for feedback here before I spend time coding it up (which is mostly adding support for adapters that I'll probably never use) and submitting a PR.  I'm already using the (provisional) `env[:typhoeus_timed_out]` custom members approach in my app and my app's custom middleware.

Does this look like a reasonable API?

**Comment by iMacTia** [comment]
Hi @nevans,
I realise I may be late on this, but I like the idea and would love to see it in the next Faraday version.
An acceptable API would be adapter-agnostic, so I would suggest we go for something like what you suggested:

```ruby
response.timed_out? #=> true/false
response.time       #=> total time in seconds
response.time_ms    #=> total time in ms
```

In order to achieve this, we'll need to add a property to `Env` (https://github.com/lostisland/faraday/blob/master/lib/faraday/options.rb#L266) which can be `execution_time` (in ms).
Then ALL adapters should populate this property upon request completion. If the adapter supports time-tracking then great, otherwise it will be a simple tracking based on `Time.now`. Not surgery-precise, but still useful 😄.
Finally, some helpers on Faraday::Response to deal with that: `time_ms`, `time`.
I don't think there will be need for `timed_out` as timeout will rise exception. I understand the issue but I think it's unrelated and should be managed in another Issue/PR

**Comment by nevans** [comment]
Thanks for the feedback.  I haven't touched this code in a long time, but I'll see if I can't fit in a few hours for it next week.

I may need to open that second PR, since the timeout exception was a major hindrance to some of my parallel code and probably my biggest motivation in digging into this in the first place. But changing that more generally is (of course) backwards incompatible (unless there's an opt-in config setting).

**Comment by iMacTia** [comment]
Thank you @nevans, that would be amazing! Appreciate you even thinking about it.
I think you can open straight away the PR for time and time_ms, wether the parallel thing can be addressed on a second PR as you suggested. In case we can't get around the backwards incompatibility, we can always decide to fit it into v1.0.
Please give me some more details about it in case you go down that route, especially since I don't have much exposure into parallel adapters like `Typhoeus`

**Comment by iMacTia** [comment]
Found this PR proposal in `faraday_middleware`: https://github.com/lostisland/faraday_middleware/issues/133

Of course an adapter-reliant value will be much more accurate, but it might be interesting to see how the middleware works and eventually take some inspiration from it
